cloud-aws-native/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/resource/instance/AwsNativeInstanceResourceBuilder.java

```diff
@@ -162,34 +162,33 @@ public List<CloudResource> update(AwsContext context, CloudInstance instance, lo
                 modifyInstanceAttributeRequest.setInstanceType(instance.getTemplate().getFlavor());
                 amazonEc2Client.modifyInstanceAttribute(modifyInstanceAttributeRequest);
             } else {
-                LOGGER.info("Instance with name: {} ({}), using the same type what was requested: {}", awsInstance.getInstanceId(),
-                        requestedInstanceType);
+                LOGGER.info("Instance ID {} using the same type what was requested: {}", awsInstance.getInstanceId(), requestedInstanceType);
             }
 
         }
         return List.of();
     }
 
-    String getSecurityGroupId(AwsContext context, Group group) {
+    @VisibleForTesting
```

Comment:
could you drop this?

Reply:
I would not drop it.
It should be `private` anyway, the visibility is loosened only for testing purposes and it is clearly seen with this annotation, as at many other places.

Reply:
you should test via public methods. I agree this should be private

Reply:
Decide which is better in the test.
```@Test
void testGetSecurityGroupIdWhenHasMoreThanOneSecurityGroupButNeedToSelectTheSecond() {
    CloudResource secGroupResource1 = getSecurityGroupResourceBuilder().withGroup("groupName1").withReference("ref1").build();
    CloudResource secGroupResource2 = getSecurityGroupResourceBuilder().withGroup("groupName2").withReference("ref2").build();

    when(group.getName()).thenReturn("groupName2");
    when(awsContext.getGroupResources("groupName2")).thenReturn(List.of(secGroupResource1, secGroupResource2));
    List<String> actual = underTest.getSecurityGroupIds(awsContext, group);
    assertThat(actual).containsExactly("ref2");



@Test
void testGetSecurityGroupIdsFromGroupReturnsAll() {
    when(group.getName()).thenReturn("groupName2");
    when(group.getSecurity()).thenReturn(security);
    when(security.getCloudSecurityIds()).thenReturn(List.of("sg1", "sg2", "sg3"));

    List<String> actual = underTest.getSecurityGroupIds(awsContext, group);
    assertThat(actual).containsExactly("sg1", "sg2", "sg3");
}
```

vs

```
@Test
void testGetSecurityGroupIdWhenHasMoreThanOneSecurityGroupButNeedToSelectTheSecond() throws Exception {
    CloudResource secGroupResource1 = getSecurityGroupResourceBuilder().withGroup("groupName1").withReference("ref1").build();
    CloudResource secGroupResource2 = getSecurityGroupResourceBuilder().withGroup("groupName2").withReference("ref2").build();

    when(group.getName()).thenReturn("groupName2");
    when(group.getReferenceInstanceTemplate()).thenReturn(instanceTemplate);
    when(awsContext.getGroupResources("groupName2")).thenReturn(List.of(secGroupResource1, secGroupResource2));
    when(awsContext.getAmazonEc2Client()).thenReturn(amazonEc2Client);
    when(awsMethodExecutor.execute(any(), any())).thenReturn(Optional.empty());
    when(awsTaggingService.prepareEc2TagSpecification(cloudStack.getTags(), com.amazonaws.services.ec2.model.ResourceType.Instance))
            .thenReturn(new TagSpecification());
    Image image = mock(Image.class);
    when(cloudStack.getImage()).thenReturn(image);
    when(image.getImageName()).thenReturn("img-name");
    InstanceAuthentication authentication = mock(InstanceAuthentication.class);
    when(cloudStack.getInstanceAuthentication()).thenReturn(authentication);
    RunInstancesResult runInstancesResult = mock(RunInstancesResult.class);
    Instance instance = new Instance().withInstanceId("instanceId");
    when(runInstancesResult.getReservation()).thenReturn(new Reservation().withInstances(instance));
    when(amazonEc2Client.createInstance(any())).thenReturn(runInstancesResult);

    underTest.build(awsContext, cloudInstance, 123L, ac, group, List.of(secGroupResource1, secGroupResource2), cloudStack);
    ArgumentCaptor<RunInstancesRequest> instanceRequestCaptor = ArgumentCaptor.forClass(RunInstancesRequest.class);
    verify(amazonEc2Client).createInstance(instanceRequestCaptor.capture());
    RunInstancesRequest actual = instanceRequestCaptor.getValue();
    assertThat(actual.getSecurityGroupIds()).containsExactly("ref2");
}


@Test
void testGetSecurityGroupIdsFromGroupReturnsAll() throws Exception {
    when(group.getName()).thenReturn("groupName2");
    when(group.getReferenceInstanceTemplate()).thenReturn(instanceTemplate);
    when(awsContext.getGroupResources("groupName2")).thenReturn(List.of());
    when(awsContext.getAmazonEc2Client()).thenReturn(amazonEc2Client);
    when(awsMethodExecutor.execute(any(), any())).thenReturn(Optional.empty());
    when(awsTaggingService.prepareEc2TagSpecification(cloudStack.getTags(), com.amazonaws.services.ec2.model.ResourceType.Instance))
            .thenReturn(new TagSpecification());
    Image image = mock(Image.class);
    when(cloudStack.getImage()).thenReturn(image);
    when(image.getImageName()).thenReturn("img-name");
    InstanceAuthentication authentication = mock(InstanceAuthentication.class);
    when(cloudStack.getInstanceAuthentication()).thenReturn(authentication);
    RunInstancesResult runInstancesResult = mock(RunInstancesResult.class);
    Instance instance = new Instance().withInstanceId("instanceId");
    when(runInstancesResult.getReservation()).thenReturn(new Reservation().withInstances(instance));
    when(amazonEc2Client.createInstance(any())).thenReturn(runInstancesResult);
    when(group.getName()).thenReturn("groupName2");
    when(group.getSecurity()).thenReturn(security);
    when(security.getCloudSecurityIds()).thenReturn(List.of("sg1", "sg2", "sg3"));
    CloudResource secGroupResource1 = getSecurityGroupResourceBuilder().withGroup("groupName1").withReference("ref1").build();

    underTest.build(awsContext, cloudInstance, 123L, ac, group, List.of(secGroupResource1), cloudStack);
    ArgumentCaptor<RunInstancesRequest> instanceRequestCaptor = ArgumentCaptor.forClass(RunInstancesRequest.class);
    verify(amazonEc2Client).createInstance(instanceRequestCaptor.capture());
    RunInstancesRequest actual = instanceRequestCaptor.getValue();
    assertThat(actual.getSecurityGroupIds()).containsExactly("sg1", "sg2", "sg3");
}

```

Reply:
Never mind, extracted the whole thing to `SecurityGroupBuilderUtil`

Reply:
the latter is better, because if I change the code on line 131 back to the old method, the first one won't fail. that's why we test the public method. if testing is too complicated then you might need to split your class into smaller ones and thats fine